### PR TITLE
add: an example on fine-tuning segformers.

### PR DESCRIPTION
**Pull request by sayakpaul**
Colab Notebook: https://colab.research.google.com/drive/1nJidirvsRgWZl2tNHFma1pArBCx87anQ?usp=sharing

Cc: @RocketKnight1 @amyeroberts

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Looks good to me! Quick question, is `ignore_mismatched_sizes=True` needed because the number of segmentation classes in this task is different from the original checkpoint?

**Comment by sayakpaul** [comment]
> Looks good to me! Quick question, is `ignore_mismatched_sizes=True` needed because the number of segmentation classes in this task is different from the original checkpoint?

I assumed so, yes.

File: examples/vision/segformer.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+* Incorporate data augmentation to potentially improve the results. 
+* Use a larger SegFormer model checkpoint to see how the results are affected. 
+* Push the fine-tuned model to the Hugging Face for sharing with the community easily.
+You can do so just by doing `model.push_to_hub("your-username/your-awesome-model")`.
+And then you can load the model by doing
+`TFSegformerForSemanticSegmentation.from_pretrained("your-username/your-awesome-model"`).
+[Here](https://github.com/huggingface/notebooks/blob/main/examples/semantic_segmentation-tf.ipynb)
+is an end-to-end example if you're looking for a reference.
+* If you'd rather want to push the model checkpoints to the Hub as the model is being
```
Comment (amyeroberts) [inline examples/vision/segformer.py:283]: nit: I'd say "If you'd rather {do task}" - don't ask me to explain the grammar rule 🙈 

```suggestion
* If you'd rather push the model checkpoints to the Hub as the model is being
```

File: examples/vision/segformer.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+"""
+Notice that we are not using any loss function for compiling the model. This is because
+the forward pass of the model
+[implements](https://github.com/huggingface/transformers/blob/820c46a707ddd033975bc3b0549eea200e64c7da/src/transformers/models/segformer/modeling_tf_segformer.py#L873)
+the loss computation part when we provide labels alongside the input images. After
+computing the loss, the model returned a structured `dataclass` object which is
+then used to guide the training process.
+
+With the compiled, we can proceed and call `fit()` on it to begin the fine-tuning
```
Comment (amyeroberts) [inline examples/vision/segformer.py:196]: nit: maybe? Or "With it compiled"? 

```suggestion
With the compiled model, we can proceed and call `fit()` on it to begin the fine-tuning
```

**Comment by amyeroberts** [comment]
@sayakpaul LGTM! Just added two ultra nits on the descriptions

**Comment by sayakpaul** [comment]
@amyeroberts thanks! You would need to sign the CLA.

File: examples/vision/segformer.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
+"""
+Title: Semantic segmentation with Hugging Face Transformers
+Author: [Sayak Paul](https://twitter.com/RisingSayak)
+Date created: 2023/01/25
+Last modified: 2023/01/25
+Description: Fine-tuning a SegFormer model variant for semantic segmentation.
+Accelerator: GPU
+"""
+"""
+# Semantic segmentation with Hugging Face Transformers
```
Comment (fchollet) [inline examples/vision/segformer.py:10]: The title line is added automatically, please remove this line

File: examples/vision/segformer.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+In this example, we will learn how to fine-tune a SegFormer model variant for the task of
+semantic segmentation on a custom dataset. SegFormer was proposed in
```
Comment (fchollet) [inline examples/vision/segformer.py:13]: Before the Segformer intro line, please add a line or two to introduce semantic segmentation

File: examples/vision/segformer.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+In this example, we will learn how to fine-tune a SegFormer model variant for the task of
+semantic segmentation on a custom dataset. SegFormer was proposed in 
+[SegFormer: Simple and Efficient Design for Semantic Segmentation with Transformers](https://arxiv.org/abs/2105.15203).
+SegFormer uses a hierarchical transformer architecture (called "Mix Transformer") as
+its encoder and a lightweight decoder for segmentation. As a result, it yields
+state-of-the-art performance on semantic segmentation while being more efficient than
+the existing models. For more details, check out the original paper. 
+
+We'll leverage
```
Comment (fchollet) [inline examples/vision/segformer.py:20]: Throughout the example, please use present tense, e.g. `we leverage` or `we use`

File: examples/vision/segformer.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
+"""
+Title: Semantic segmentation with Hugging Face Transformers
```
Comment (fchollet) [inline examples/vision/segformer.py:2]: Please add a mention of SegFormer in the title

**Review by fchollet — COMMENTED** [review]
Thank you for the PR -- it looks great! Clear and to the point :)

**Comment by sayakpaul** [comment]
@fchollet thanks!

I have applied your suggestions and have also added an architecture diagram. PTAL.

**Review by fchollet — COMMENTED** [review]
Thanks for the update -- everything looks great! I pushed some minor copyedits. You can add the generated files now.

**Comment by sayakpaul** [comment]
@fchollet thanks! Added the generated files. 

Over to you.

**Review by fchollet — APPROVED** [review]
Thank you! 🤗

**Pull request merged**